File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root:
            return None

        if self.is_ancestor(root.left, p) and self.is_ancestor(root.left, q):
            return self.lowestCommonAncestor(root.left, p, q)

        if self.is_ancestor(root.right, p) and self.is_ancestor(root.right, q):
            return self.lowestCommonAncestor(root.right, p, q)

        if self.is_ancestor(root, p) and self.is_ancestor(root, q):
            return root
        
    def is_ancestor(self, root, node) -> bool:
        if not root:
            return False

        if root.val == node.val:
            return True

        return self.is_ancestor(root.left, node) or self.is_ancestor(root.right, node)

File: Data Structures & Algorithms/binary-tree-vertical-order-traversal/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 * 
 * interface NodeMetadata {
 *      row: number;
 *      column: number;
 *      node: TreeNode
 * }
 */



class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    verticalOrder(root) {
        if (!root) return [];
        const colToNodeAndRow = new Map(); // key: column, value: [Node, Row]

        function levelOrderTraversal(root) {
            let [leftmostCol, rightmostCol] = [0,0];
            if (!root) return [leftmostCol, rightmostCol];
            const queue = [];
            let queueItem = {
                node: root,
                row: 0,
                col: 0
            }
            queue.push(queueItem);
            while (queue.length !== 0) {
                const { node: currentNode, row: currentRow, col: currentCol } = queue.shift();
                if (!currentNode) continue;

                if (currentNode.left) {
                    const nextNode = {
                        node: currentNode.left,
                        row: currentRow+1,
                        col: currentCol-1
                    }
                    queue.push(nextNode);
                    if (currentCol-1 < leftmostCol) leftmostCol = currentCol-1;
                }
                if (currentNode.right) {
                    const nextNode = {
                        node: currentNode.right,
                        row: currentRow+1,
                        col: currentCol+1
                    }
                    queue.push(nextNode);
                    if (currentCol+1 > rightmostCol) rightmostCol = currentCol+1;
                }

                if (colToNodeAndRow.has(currentCol)) {
                    const updated = colToNodeAndRow.get(currentCol);
                    updated.push({node: currentNode, row: currentRow});
                    colToNodeAndRow.set(currentCol, updated);
                } else {
                    colToNodeAndRow.set(currentCol, [{node: currentNode, row: currentRow}]);
                }

            }
            return [leftmostCol, rightmostCol];
        }

        const [leftmostCol, rightmostCol] = levelOrderTraversal(root);
        const result = [];
        for (let i = leftmostCol; i <= rightmostCol; i++) {
            const currentCol = colToNodeAndRow.get(i);
            const vals = currentCol.map((cell) => cell.node.val);
            // represents current array of [node:row] elements
            // colToNodeAndRow[i].sort(([a,b]))
            result.push(vals);
        }

        return result;
        
    }
}







